src/transports/tcp.ts

```diff
@@ -0,0 +1,236 @@
+                    this.openStream();
+                });
+                
+                const handleFeatures = (features: StreamFeatures) => {
+                    if (!('tls' in features)) {
+                        if (this.client.config.requireSecureTransport && !this.isSecure) {
+                            this.client.emit('--transport-error', new Error('failed to connect - STARTLS not offered'));
+                            this.client.off('features', handleFeatures);
+                        } else {
+                            this.client.emit('stream:start', {});
+                            this.client.emit('--transport-connected');
+                            this.client.off('features', handleFeatures);
+                        }
+                    }
+                };
+
+                this.client.on('features', handleFeatures);
+                this.client.once('--transport-connected', () => resolve());
+                this.client.once('--transport-error', e => reject(e));
+                this.socket.on('data', packet => this.parser!.write(packet.toString('utf8')));
+            }
+        });
+    }
+    
+    public _write(chunk: string, encoding: string, done: (err?: Error) => void): void {
+        const data = Buffer.from(chunk, 'utf8').toString();
+        this.client.emit('raw', 'outgoing', data);
+        (this.tlssocket || this.socket)?.write(data);
+        done();
+    }
+    
+    public _read(): void { return; }
+    
+    public disconnect(cleanly = true): void {
+        if (cleanly) this.write(`</stream:stream>`);
+        setTimeout(() => {
+            this.hasStream = false;
+            (this.tlssocket || this.socket)?.destroy();
+            this.emit('end');
+        }, cleanly ? 500 : 0);
+    }
+    
+    public restart(): void {
+        this.hasStream = false;
+        this.initParser();
+        this.openStream();
+    }
+    
+    public async send(name: string, data?: JSONData): Promise<void> {
+        let output: string | undefined;
+        if (data) {
+            output = this.stanzas.export(name, data)?.toString();
+        }
+        if (!output) {
+            return;
+        }
+        
+        return new Promise<void>((resolve, reject) => {
+            this.write(output, 'utf8', err => (err ? reject(err) : resolve()));
+        });
+    }
+    
+    private openStream(): void {
+        this.write(`<?xml version='1.0'?><stream:stream from='${this.config.jid}' to='${this.config.server}' version='1.0' xml:lang='${this.config.lang}' xmlns='jabber:client' xmlns:stream='http://etherx.jabber.org/streams'>`);
+    }
+
+    private async negotiateTls(): Promise<void> {
+        return await new Promise<void>((resolve, reject) => {
+            this.tlssocket = tls.connect({
+                socket: this.socket!,
+                checkServerIdentity: (host: string, cert: tls.PeerCertificate): Error | undefined => {
+                    if (!this.config.pubkey) return undefined;
```

Comment:
Wait, so you have to manually specify the public key? What about using system approved certs and CAs?

Reply:
You don't have to specify the public key at all, it's optional in `TransportConfig`. I do agree though, we should definitely be checking certs against the OS, that would make it much more browser-like in its security. I'll look into how to perform that kind of check.

Reply:
Node TLS already supports a predefined list of CAs curated by Mozilla. By default these are the CAs used:

> rejectUnauthorized `<boolean>` If not false, the server certificate is verified against the list of supplied CAs. An 'error' event is emitted if verification fails; err.code contains the OpenSSL error code. Default: `true`.

You can also provide your own CA list either by passing a `secureContext` with its `ca` option set or by directly providing a `ca` option to the opts in `connect` or TLSSocket's constructor.

Reply:
The documentation on this should be pristinely clear, maybe changing publicly-facing `cert` to `customCert`, etc. etc.
If Mozilla's CA list is long enough to where it practically won't be an issue, this is a lower prio issue. Stick a `TODO` and/or a new issue.